cluster/src/Cardano/Cluster/Environment.hs

```diff
@@ -254,64 +265,39 @@ prepareEnvironment node@(NodeName nodeIdT, nodeType) nodes stateDir = runState $
     prepareTopology :: Env -> (Artifact Topology (), Env)
     prepareTopology env =
         let
-            cIndex :: Int
-            cIndex =
-                unsafeElemIndex node nodes
-
-            addr :: NetworkAddress
-            addr =
-                -- NOTE Safe when called after 'withDefaultEnvironment'
-                unsafeNetworkAddressFromString (env ! "LISTEN")
-
-            waddr :: NetworkAddress
-            waddr =
-                -- NOTE Safe when called after 'withDefaultEnvironment'
-                unsafeNetworkAddressFromString (env ! "NODE_API_ADDRESS")
-
             topologyPath :: FilePath
             topologyPath =
                 stateDir </> "topology" </> T.unpack nodeIdT <> ".json"
 
-            (nodeNames :: [NodeName], nodeTypes :: [NodeType]) =
-                unzip nodes
-
-            nodeAddrs :: [NetworkAddress]
-            nodeAddrs = flip evalState (0, 0, 0) $ forM nodeTypes $ \typ -> do
-                (c, r, w) <- get
-                case typ of
-                    NodeCore ->
-                        put (c + 1, r, w) >> return (nextNtwrkAddr c addr)
-                    NodeRelay ->
-                        put (c, r + 1, w) >> return (nextNtwrkAddr (r + 100) addr)
-                    NodeEdge ->
-                        put (c, r, w + 1) >> return (nextNtwrkAddr w waddr)
+            (listenAddrs :: [Maybe NetworkAddress], apiAddrs :: [NetworkAddress]) =
+                unzip $ nodeAddrs env
 
             topology :: Topology
             topology =
-                demoTopology nodeType (zip3 nodeNames nodeTypes nodeAddrs)
+                case nodeType of
+                    NodeEdge -> demoTopologyBehindNAT
+                        $ map (\(a, _, Just c) -> (a, c)) -- Safe, no edges
+                        $ filter isRelayNode
+                        $ zip3 nodeNames nodeTypes listenAddrs
+
+                    _ -> demoTopologyStatic
+                        $ map (\(a, b, Just c) -> (a, b, c)) -- Safe, no edges
+                        $ filter (not . isEdgeNode)
+                        $ zip3 nodeNames nodeTypes listenAddrs
```

Comment:
I've also split `demoTopology` into two functions with slightly different arguments, instead of conflating both with a `case nodeType of` inside `demoTopology. It's a bit clearer.